File: Xenon/Xenon/OrganicGun.ts
import CActor = require("Actor")
import gsCControls = require("Controls")
import enums = require("Enums");
import CSpinnerWeapon = require("SpinnerWeapon");
import gsCVector = require("Vector");
import CPlayGameState = require("PlayGameState");
import CExplode = require("Exploder");

//-------------------------------------------------------------

enum OrganicGunState {
    ORGANICGUN_STILL,
    ORGANICGUN_SHOOTING,
}

//-------------------------------------------------------------

class COrganicGun extends CActor {

    ORGANICGUN_SHOT_START: number = 0;
    ORGANICGUN_SHOT_FRAMES: number = 8;
    ORGANICGUN_LAUNCH_FRAME: number = 6;
    ORGANICGUN_LEFT: number = 0;
    ORGANICGUN_RIGHT: number = 8;
    ORGANICGUN_STILL_TIME: number = 1.0;

    private m_weapon: CSpinnerWeapon;
    private m_side: number;
    private m_direction: number;
    private m_fired: boolean;
    private m_state: OrganicGunState;
    private m_random;//static gsCRandom 

    //-------------------------------------------------------------

    constructor(playGameState: CPlayGameState) {
        super();
        this.m_weapon = null;
        this.m_fired = false;
        this.m_name = "OrganicGun";
        this.m_playGameState = playGameState;
    }

    //-------------------------------------------------------------

    public getActorInfo() {
        this.m_actorInfo = this.m_scene.GetlistOfActors();
        return this.m_actorInfo.GetActorInfoListItem(enums.ActorInfoType.INFO_ORGANIC_GUN);
    }

    //-------------------------------------------------------------

    public activate(): boolean {
        if (!this.isActive()) {
            this.m_weapon = new CSpinnerWeapon();
            this.m_scene.addActor(this.m_weapon);
            this.m_weapon.activate();
            this.m_weapon.setOwner(this);
            this.m_weapon.setOffset(new gsCVector(16.0, 16.0));
            this.m_weapon.setFiringMode(enums.WeaponFiringMode.WEAPON_MANUAL);
            this.m_weapon.setGrade(enums.WeaponGrade.WEAPON_BEST);

            this.m_state = OrganicGunState.ORGANICGUN_STILL;
            this.m_timer.start();
        }

        return super.activate();
    }

    //-------------------------------------------------------------

    public kill(): void {
        if (this.m_weapon) {
            this.m_weapon.kill();
            this.m_weapon = null;
        }
        super.kill();
    }

    //-------------------------------------------------------------

    public setDirection(dir: number): void {
        if (dir > 0) {
            this.m_side = this.ORGANICGUN_LEFT;
            this.m_weapon.setDirectionS(new gsCVector(1.0, 0.0));
        }
        else {
            this.m_side = this.ORGANICGUN_RIGHT;
            this.m_weapon.setDirectionS(new gsCVector(-1.0, 0.0));
        }
    }

    //-------------------------------------------------------------

    public update(controls: gsCControls): boolean {
        if (this.m_shield == 0) {
            var explode = new CExplode(this);
            this.kill();
            return true;
        }

        switch (this.m_state) {
            case OrganicGunState.ORGANICGUN_STILL:
                this.m_sprite.setFrame(this.m_side + this.ORGANICGUN_SHOT_START);
                if (this.m_timer.getTime() >= this.ORGANICGUN_STILL_TIME) {

                    this.m_state = OrganicGunState.ORGANICGUN_SHOOTING;
                    this.m_fired = false;
                    this.m_timer.start();
                }
                break;

            case OrganicGunState.ORGANICGUN_SHOOTING:
                var frame: number = this.m_timer.getTime() * this.getActorInfo().m_anim_rate;
                if (frame >= this.ORGANICGUN_SHOT_FRAMES) {
                    this.m_sprite.setFrame(this.m_side + this.ORGANICGUN_SHOT_START);
                    this.m_state = OrganicGunState.ORGANICGUN_STILL;
                    this.m_timer.start();
                }
                else {
                    this.m_sprite.setFrame(this.m_side + this.ORGANICGUN_SHOT_START + frame);
                    if (!this.m_fired && frame >= this.ORGANICGUN_LAUNCH_FRAME) {
                        this.m_weapon.fire();
                        this.m_fired = true;
                    }
                }
                break;
        }
        return true;
    }

    //-------------------------------------------------------------

    public onLeavingScreen(): void {
        if (!this.m_playGameState.reachedBoss()) {
            this.kill();
        }
    }

    //-------------------------------------------------------------

}
export = COrganicGun;

File: Xenon-Original_C++_Code/xenon/source/organicgun.cpp
//-------------------------------------------------------------
//
// Class:	COrganicGun
//
// Author:	John M Phillips
//
// Started:	06/05/00
//
// Base:	CAlien
//
// Derived:	None
//
//-------------------------------------------------------------

#include "game.h"

//-------------------------------------------------------------

gsCRandom COrganicGun::m_random;

//-------------------------------------------------------------

COrganicGun::COrganicGun()
{
	m_weapon = 0;
	m_fired = false;
}

//-------------------------------------------------------------

COrganicGun::~COrganicGun()
{
}

//-------------------------------------------------------------

void COrganicGun::setDirection(int dir)
{
	if (dir > 0) {
		m_side = ORGANICGUN_LEFT;
		m_weapon->setDirection(gsCVector(1.f,0.f));
		}
	else {
		m_side = ORGANICGUN_RIGHT;
		m_weapon->setDirection(gsCVector(-1.f,0.f));
		}
}

//-------------------------------------------------------------

bool COrganicGun::activate()
{
	if (!isActive()) {
		m_weapon = new CSpinnerWeapon;
		m_scene->addActor(m_weapon);
		m_weapon->activate();
		m_weapon->setOwner(this);
		m_weapon->setOffset(gsCVector(16.f,16.f));
		m_weapon->setFiringMode(WEAPON_MANUAL);
		m_weapon->setGrade(WEAPON_BEST);

		m_state = ORGANICGUN_STILL;
		m_timer.start();
		}

	return CActor::activate();
}

//-------------------------------------------------------------
void COrganicGun::kill()
{
	if (m_weapon) {
		m_weapon->kill();
		m_weapon = 0;
		}

	CActor::kill();
}

//-------------------------------------------------------------
		
bool COrganicGun::update(Controls *controls)
{
	if (m_shield == 0) {
		explode();
		kill();
		return true;
		}

	switch (m_state) {
		case ORGANICGUN_STILL:
			{
			m_sprite.setFrame(m_side + ORGANICGUN_SHOT_START);

			if (m_timer.getTime() >= ORGANICGUN_STILL_TIME) {

				m_state = ORGANICGUN_SHOOTING;
				m_fired = false;
				m_timer.start();
				}
			}
			break;
		case ORGANICGUN_SHOOTING:
			{
				int frame = (int) (m_timer.getTime() * getActorInfo().m_anim_rate);
				if (frame >= ORGANICGUN_SHOT_FRAMES) {
					m_sprite.setFrame(m_side + ORGANICGUN_SHOT_START);
					m_state = ORGANICGUN_STILL;
					m_timer.start();
					}			
				else {
					m_sprite.setFrame(m_side + ORGANICGUN_SHOT_START + frame);
					if (!m_fired && frame >= ORGANICGUN_LAUNCH_FRAME) {
						m_weapon->fire();
						m_fired = true;
						}
					}
			}
			break;
		}

	return true;
}

//-------------------------------------------------------------

void COrganicGun::onLeavingScreen()
{
	if (!CPlayGameState::reachedBoss())
		kill();
}

//-------------------------------------------------------------

